### PR TITLE
Some work related to countable types

### PR DESCRIPTION
**Pull request by dolio**
This PR contains an assortment of stuff written working up to definitions of countability and similar properties.

Miscellaneous convenience stuff:
- Discreteness can be translated through a section
- `isSet` for plain functions when the codomain is a set
- `isPropDep` alias
- Generalized the type of `PropositionalTruncation.rec2`

More properties of the recursively defined ordering on the natural numbers, including well-foundedness.

Define various omniscience principles in `Cubical.Axiom.Omniscience`, along with how they imply one another.

Gives some more specific facts about omniscience on the naturals (could probably be generalized to any well-founded total ordering).

Define some additional structure on the `Conat` type, including the type of natural numbers bounded by a conat, which is analogous to `Fin`. Prove that this is similarly injective.

Show that an injective function from `Conat` to `ℕ` implies the weak limited principle of omniscience on the naturals. Show that a certain pseudo 'covering' function from `ℕ` to `Conat` being an equivalence implies the limited principle of omniscience.

Define various notions of countability, their interaction, and displayed versions:
- Denumerability for being merely equivalent to `ℕ`
- Countability for being equivalent to a set with a `Conat` size
- CountableIndexability for admitting a surjective 'partial' function from the naturals
These are analogous to notions of being finite.

### CODE REVIEW

**Comment by dolio** [comment]
There seems to be an issue where the `Everything` file generation doesn't include `--guardedness` for files that transitively import files that use it, causing tests to fail. Not certain what to do about that.

**Comment by dolio** [comment]
I hit more snags than I expected, but I was able to reprove the countability stuff for detachable lower segments of ℕ, so I switched to that, since it seems like less of a hassle tooling wise.

**Comment by Saizan** [comment]
Any idea what's going wrong with the test?

**Comment by dolio** [comment]
I think it's just complaining that some modules are not transitively imported by `Everything` modules because I switched the basis for countability away from coinduction. Sorry about the delay fixing it. I can't easily see the issue locally, because for some reason my local agda version instead reports a ton of unsolved metas in `ZCohomology`.

**Comment by ecavallo** [comment]
Could the unsolved metas be related to #508? In that case merging from master should take care of them.

**Comment by dolio** [comment]
Oh, possibly. I guess my branch is kind of behind. Thanks for the tip.

**Comment by dolio** [comment]
Okay, I'm not quite sure what the problem is now. It seems to be in a module I didn't modify, and it also doesn't happen to me locally. Possibly it's specific to the version of Agda used?

**Comment by ecavallo** [comment]
Yes, another problem due to changes in a new version of Agda. I fixed this in #518, so merge from master again :sweat_smile:

**Review by ecavallo — CHANGES_REQUESTED** [review]
Looking good! I have a few requests and suggestions before we merge.

File: Cubical/Axiom/Omniscience.agda [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+LLPO A
+  = ∀(P Q : A → 𝟚)
+  → (∀ x y → ⟨ P x ⟩ → ⟨ Q y ⟩ → ⊥)
+  → ∥ (∀ x → ¬ ⟨ P x ⟩) ⊎ (∀ y → ¬ ⟨ Q y ⟩) ∥
+
+LLPO-isProp : isProp (LLPO A)
+LLPO-isProp = isPropΠ3 λ _ _ _ → squash
+
+-- As above, but without ensuring propositionality
+LLPO' : Type ℓ → Type ℓ
```
[the remark this answers is not in the document]
Comment (dolio) [reply Cubical/Axiom/Omniscience.agda:39]: Sure, I can go with that. Although it's never more than a set.

File: Cubical/Data/Bool/Properties.agda [hunk 2]
```diff
@@ -193,3 +195,30 @@ module BoolReflection where
...
   reflectEquiv : Bool ≃ (Bool ≡ Bool)
   reflectEquiv = isoToEquiv reflectIso
+
+_≤_ : Bool → Bool → Type
+true ≤ false = ⊥
+_ ≤ _ = Unit
+
+_≥_ : Bool → Bool → Type
+false ≥ true = ⊥
+_ ≥ _ = Unit
```
[the remark this answers is not in the document]
Comment (dolio) [reply Cubical/Data/Bool/Properties.agda:205]: This one is due to the exact behavior of reductions being important to the simplicity of other proofs. Just using `c ≤ b` gets different judgmental equality.

File: Cubical/Data/Nat/Lower.agda [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+mpred f .snd k = f .snd (suc k)
+
+data MView (m : ℕ → Bool) : Type where
+  mzv : mz ≡ m → MView m
+  msv : ∀ n → ms n ≡ m → MView m
+
+mview : ∀ f → isMono f → MView f
+mview f mf with f 0 | inspect f 0
+... |  true | [ p ]ᵢ = msv (mp f) (ms-mp f p)
+... | false | [ p ]ᵢ = mzv λ i k → mz-lemma f mf p k i
```
[the remark this answers is not in the document]
Comment (dolio) [reply Cubical/Data/Nat/Lower.agda:86]: Okay, sure. I guess I was eagerly avoiding indexed types due to them (currently) getting stuck on some cubical stuff, but it probably doesn't matter in this example.

File: Cubical/Displayed/Countable.agda [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+open import Cubical.Displayed.Base
+open import Cubical.Displayed.Properties
+
+private
+  variable
+    ℓ : Level
+    A : Type ℓ
+
+Denumeration : Type ℓ → Type ℓ
```
[the remark this answers is not in the document]
Comment (dolio) [reply Cubical/Displayed/Countable.agda:33]: I'm fine to move them. I guess I was thinking forward to synthetic computability axioms, like, 'the countably enumerable subsets of ℕ are countably enumerable,' and thinking the 'structure' bits might be used for that. But I wasn't super convinced of their utility myself, really.

File: Cubical/Data/Bool/Properties.agda (hunk 2 above)
[the remark this answers is not in the document]
Comment (ecavallo) [reply Cubical/Data/Bool/Properties.agda:205]: Oh, that makes sense.

**Pull request merged**